pipelines/common/src/main/java/org/openmrs/analytics/DwhFiles.java

```diff
@@ -154,39 +178,96 @@ public Set<String> findResourceTypes() throws IOException {
     return typeSet;
   }
 
+  /**
+   * This method copies all the files under the directory (getRoot() + resourceType) into the
+   * destination DwhFiles under the similar directory.
+   *
+   * @param resourceType
+   * @param destDwh
+   * @throws IOException
+   */
   public void copyResourcesToDwh(String resourceType, DwhFiles destDwh) throws IOException {
-    Path path = fileSystem.getPath(getRoot(), resourceType);
-    Path destPath = destDwh.getResourcePath(resourceType);
-    log.info("Copying {} to {}", path, destPath);
-    FileUtils.copyDirectory(path.toFile(), destPath.toFile());
+    ResourceId sourceFiles =
+        FileSystems.matchNewResource(getRoot(), true)
+            .resolve(resourceType, StandardResolveOptions.RESOLVE_DIRECTORY)
+            .resolve("*", StandardResolveOptions.RESOLVE_FILE);
+
+    List<MatchResult> sourceMatchResultList = FileSystems.matchResources(List.of(sourceFiles));
+    List<ResourceId> sourceResourceIdList = new ArrayList<>();
+    for (MatchResult matchResult : sourceMatchResultList) {
+      if (matchResult.status() == Status.OK) {
+        List<ResourceId> resourceIds =
+            matchResult.metadata().stream()
+                .map(metadata -> metadata.resourceId())
+                .collect(Collectors.toList());
+        sourceResourceIdList.addAll(resourceIds);
+      }
+    }
+
+    List<ResourceId> destResourceIdList = new ArrayList<>();
+    sourceResourceIdList.stream()
+        .forEach(
+            resourceId -> {
+              ResourceId destResourceId =
+                  FileSystems.matchNewResource(destDwh.getRoot(), true)
+                      .resolve(resourceType, StandardResolveOptions.RESOLVE_DIRECTORY)
+                      .resolve(resourceId.getFilename(), StandardResolveOptions.RESOLVE_FILE);
+              destResourceIdList.add(destResourceId);
+            });
+    FileSystems.copy(sourceResourceIdList, destResourceIdList);
```

Comment:
Do you know if this uses the parallel mode of GCS copy? I know that when doing copy from the command-line, there is a `-m` option (as described [here](https://cloud.google.com/storage/docs/gsutil/commands/cp)) and the performance can be significantly different.

Reply:
Here the copy is made in batches for the GcsFileSystem, where each batch contain 100 files in max to copy. The batches are executed concurrently using the ExecutorService. However, for the LocalFileSystem the copy is sequential and all the files are copied one after the other.